internal/tar_ball_queue.go

```diff
@@ -57,11 +58,22 @@ func (tarQueue *TarBallQueue) StartQueue() error {
 	return nil
 }
 
-func (tarQueue *TarBallQueue) Deque() TarBall {
+// DequeCtx returns a TarBall from the queue. If the context finishes before it
+// can do so, then instead it returns nil.
+func (tarQueue *TarBallQueue) DequeCtx(ctx context.Context) TarBall {
 	if tarQueue.started.IsNotSet() {
 		panic("Trying to deque from not started Queue")
 	}
-	return <-tarQueue.tarsToFillQueue
+	select {
+	case <-ctx.Done():
+		return nil
```

Comment:
Last thing:
In case of some weird situation when there is a nil tarball in the tarQueue.tarsToFillQueue, it might be hard to distinguish it from canceled context.

So I suggest returning an explicit error instead:
```
return nil, ctx.Err()
```

Reply:
Done